File: StampService/src/StampService.Domain/User/PhoneNumber.cs
using FluentResults;
using StampService.Domain.Shared;

namespace StampService.Domain.User;

public sealed record PhoneNumber
{
    public const int MinDigitsLength = 2;
    public const int MaxDigitsLength = 15;
    public const int MaxInputLength = 32;
    public const int CountryCodeSevenDigitsLength = 11;

    private PhoneNumber(string value)
    {
        Value = value;
    }

    public string Value { get; }

    public static Result<PhoneNumber> Create(string? input)
    {
        var normalizedResult = Normalize(input);
        if (normalizedResult.IsFailed)
            return Result.Fail(normalizedResult.Errors);

        return Result.Ok(new PhoneNumber(normalizedResult.Value));
    }

    public static Result<string> Normalize(string? input)
    {
        if (string.IsNullOrWhiteSpace(input))
            return Invalid(nameof(input));

        var trimmed = input.Trim();
        if (trimmed.Length > MaxInputLength)
            return Invalid(nameof(input));

        if (trimmed[0] != '+')
            return Invalid(nameof(input));

        if (trimmed.Count(character => character == '+') != 1)
            return Invalid(nameof(input));

        if (trimmed.Skip(1).Any(character => !IsAllowedAfterPlus(character)))
            return Invalid(nameof(input));

        var digits = new string(trimmed.Where(IsAsciiDigit).ToArray());
        if (digits.Length is < MinDigitsLength or > MaxDigitsLength)
            return Invalid(nameof(input));

        if (digits[0] == '0')
            return Invalid(nameof(input));

        if (digits[0] == '7' && digits.Length != CountryCodeSevenDigitsLength)
            return Invalid(nameof(input));

        return Result.Ok($"+{digits}");
    }

    public static bool IsValidNormalized(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return false;

        if (value.Length > MaxInputLength || value[0] != '+')
            return false;

        var digits = value[1..];
        if (digits.Length is < MinDigitsLength or > MaxDigitsLength)
            return false;

        if (!digits.All(IsAsciiDigit) || digits[0] == '0')
            return false;

        return digits[0] != '7' || digits.Length == CountryCodeSevenDigitsLength;
    }

    public override string ToString() => Value;

    private static bool IsAllowedAfterPlus(char character)
    {
        return IsAsciiDigit(character)
            || character == ' '
            || character == '-'
            || character == '('
            || character == ')';
    }

    private static bool IsAsciiDigit(char character) => character is >= '0' and <= '9';

    private static Result<string> Invalid(string field)
    {
        return Result.Fail(DomainError.Validation(
            "phone_number.invalid",
            "Phone number is invalid",
            field));
    }
}

File: StampService/src/StampService.Web/src/validation/phoneNumber.ts
const minDigitsLength = 2;
const maxDigitsLength = 15;
const maxInputLength = 32;
const countryCodeSevenDigitsLength = 11;

export type PhoneNumberValidationResult =
  | { ok: true; value: string }
  | { ok: false; message: string };

export function formatRuPhoneInput(input: string): string {
  const localDigits = getRuLocalDigits(input);
  const operator = localDigits.slice(0, 3);
  const firstPart = localDigits.slice(3, 6);
  const secondPart = localDigits.slice(6, 8);
  const thirdPart = localDigits.slice(8, 10);

  let value = '+7';

  if (operator) {
    value += ` (${operator}`;
  }

  if (operator.length === 3) {
    value += ')';
  }

  if (firstPart) {
    value += ` ${firstPart}`;
  }

  if (secondPart) {
    value += `-${secondPart}`;
  }

  if (thirdPart) {
    value += `-${thirdPart}`;
  }

  return value;
}

export function isRuPhoneInputComplete(input: string): boolean {
  return getRuLocalDigits(input).length === 10;
}

export function normalizePhoneNumber(input: string): PhoneNumberValidationResult {
  const trimmed = input.trim();
  if (!trimmed) {
    return invalid();
  }

  if (trimmed.length > maxInputLength) {
    return invalid();
  }

  if (!trimmed.startsWith('+')) {
    return invalid();
  }

  if (countPlus(trimmed) !== 1) {
    return invalid();
  }

  for (const character of trimmed.slice(1)) {
    if (!isAllowedAfterPlus(character)) {
      return invalid();
    }
  }

  const digits = Array.from(trimmed).filter(isAsciiDigit).join('');
  if (digits.length < minDigitsLength || digits.length > maxDigitsLength) {
    return invalid();
  }

  if (digits[0] === '0') {
    return invalid();
  }

  if (digits[0] === '7' && digits.length !== countryCodeSevenDigitsLength) {
    return invalid();
  }

  return { ok: true, value: `+${digits}` };
}

function countPlus(value: string): number {
  return Array.from(value).filter((character) => character === '+').length;
}

function isAllowedAfterPlus(character: string): boolean {
  return isAsciiDigit(character)
    || character === ' '
    || character === '-'
    || character === '('
    || character === ')';
}

function isAsciiDigit(character: string): boolean {
  return character >= '0' && character <= '9';
}

function getRuLocalDigits(input: string): string {
  const trimmed = input.trim();
  const digits = Array.from(input).filter(isAsciiDigit).join('');
  if (!digits) {
    return '';
  }

  let localDigits = digits;
  if (trimmed.startsWith('+7') && localDigits[0] === '7') {
    localDigits = localDigits.slice(1);
  } else if ((localDigits[0] === '7' || localDigits[0] === '8') && localDigits.length > 10) {
    localDigits = localDigits.slice(1);
  }

  while (localDigits[0] === '7' || localDigits[0] === '8') {
    localDigits = localDigits.slice(1);
  }

  return localDigits.slice(0, 10);
}

function invalid(): PhoneNumberValidationResult {
  return {
    ok: false,
    message: 'Введите телефон в формате +7 (999) 123-45-67.'
  };
}
